test/e2e/kubevirt.go

```diff
@@ -524,7 +499,104 @@ passwd:
+				vmCreationRetries++
+				return err
+			}).WithPolling(time.Second).WithTimeout(time.Minute).Should(Succeed())
+		}
+
+		createVirtualMachineInstance = func(vmi *kubevirtv1.VirtualMachineInstance) {
+			By(fmt.Sprintf("Create virtual machine instance %s", vmi.Name))
+			vmiCreationRetries := 0
+			Eventually(func() error {
+				if vmiCreationRetries > 0 {
+					// retry due to unknown issue where kubevirt webhook gets stuck reading the request body
+					// https://github.com/ovn-org/ovn-kubernetes/issues/3902#issuecomment-1750257559
+					By(fmt.Sprintf("Retrying vmi %s creation", vmi.Name))
+				}
+				err := crClient.Create(context.Background(), vmi)
+				vmiCreationRetries++
+				return err
+			}).WithPolling(time.Second).WithTimeout(time.Minute).Should(Succeed())
+		}
+
+		waitVirtualMachineInstanceReadiness = func(vmi *kubevirtv1.VirtualMachineInstance) {
+			By(fmt.Sprintf("Waiting for readiness at virtual machine %s", vmi.Name))
+			Eventually(func() []kubevirtv1.VirtualMachineInstanceCondition {
+				err := crClient.Get(context.Background(), crclient.ObjectKeyFromObject(vmi), vmi)
+				Expect(err).To(SatisfyAny(WithTransform(apierrors.IsNotFound, BeTrue()), Succeed()))
+				return vmi.Status.Conditions
+			}).WithPolling(time.Second).WithTimeout(5 * time.Minute).Should(ContainElement(SatisfyAll(
+				HaveField("Type", kubevirtv1.VirtualMachineInstanceReady),
+				HaveField("Status", corev1.ConditionTrue),
+			)))
+		}
+
+		waitVirtualMachineAddresses = func(vmi *kubevirtv1.VirtualMachineInstance) []kubevirt.Address {
+			step := by(vmi.Name, "Wait for virtual machine to receive IPv4 address from DHCP")
+			Eventually(addressByFamily(ipv4, vmi)).
+				WithPolling(time.Second).
+				WithTimeout(5*time.Minute).
+				Should(HaveLen(1), step)
+			addresses, err := addressByFamily(ipv4, vmi)()
+			Expect(err).ToNot(HaveOccurred())
+			if isDualStack() {
+				output, err := kubevirt.RunCommand(vmi, `echo '{"interfaces":[{"name":"enp1s0","type":"ethernet","state":"up","ipv4":{"enabled":true,"dhcp":true},"ipv6":{"enabled":true,"dhcp":true,"autoconf":false}}],"routes":{"config":[{"destination":"::/0","next-hop-interface":"enp1s0","next-hop-address":"fe80::1"}]}}' |nmstatectl apply`, 5*time.Second)
+				Expect(err).ToNot(HaveOccurred(), output)
+				step = by(vmi.Name, "Wait for virtual machine to receive IPv6 address from DHCP")
+				Eventually(addressByFamily(ipv6, vmi)).
+					WithPolling(time.Second).
+					WithTimeout(5*time.Minute).
+					Should(HaveLen(2), func() string {
+						output, _ := kubevirt.RunCommand(vmi, "journalctl -u nmstate", 2*time.Second)
+						return step + " -> journal nmstate: " + output
+					})
+				ipv6Addresses, err := addressByFamily(ipv6, vmi)()
+				Expect(err).ToNot(HaveOccurred())
+				addresses = append(addresses, ipv6Addresses...)
+			}
+			return addresses
+		}
+
+		waitVirtualMachineAddressesFromStatus = func(vmi *kubevirtv1.VirtualMachineInstance, expectedNumberOfAddresses int) []string {
```

Comment:
would it make sense instead to pass a variadic number of subnets and for each of them assert there's an IP in the VM status in that range ?

Reply:
well the VMI we are configuring only expose the interface from the nad, the fact that those IPs belong to the nad.subnet is already tested at multihoming, what we want to test here is that the IPs whatever they are should prevail.

Reply:
yes, but by doing that, the function's signature looks a lot better IMO.